File: motion_decision.py
import time
import cv2
import torch
import numpy as np
import segmentation_models_pytorch as smp
from collections import defaultdict
import matplotlib.pyplot as plt

class GroundSegmenter:
    def __init__(self, ckpt_path, device='cpu'):
        self.device = torch.device(device)
        self.model = torch.jit.load(ckpt_path, map_location=self.device)
        self.model.eval()

    def predict_mask(self, img_bgr):
        h, w = img_bgr.shape[:2]
        resized = cv2.resize(img_bgr, (320, 320))
        rgb = cv2.cvtColor(resized, cv2.COLOR_BGR2RGB)
        tensor = torch.from_numpy(rgb).permute(2, 0, 1).unsqueeze(0).float() / 255.0
        tensor = tensor.to(self.device)
        with torch.no_grad():
            output = self.model(tensor).squeeze().cpu().numpy()
        mask = (output > 0.5).astype(np.uint8) * 255
        return cv2.resize(mask, (w, h))

class MotionDeciderWithLasers:
    def __init__(self, segmenter, img_width, img_height, num_beams=5):
        self.segmenter = segmenter
        self.width = img_width
        self.height = img_height
        self.num_beams = num_beams
        angle_step = 180 // (self.num_beams - 1)
        self.regions = [str(-90 + i * angle_step) for i in range(self.num_beams)]

    def laser_scan(self, mask):
        laser_scores = defaultdict(int)
        step = self.width // self.num_beams
        for i, region in enumerate(self.regions):
            x_start = i * step
            x_end = x_start + step
            region_mask = mask[self.height//2:, x_start:x_end]  # bottom half only
            laser_scores[region] = np.sum(region_mask == 255)
        return laser_scores

    def decide(self, laser_scores, threshold_ratio=0.05):
        total = sum(laser_scores.values())
        if total == 0:
            return "stop"
        best_region = max(laser_scores.items(), key=lambda x: x[1])[0]
        if laser_scores[best_region] < total * threshold_ratio:
            return "stop"
        return best_region

def process_image(image_path, segmenter, decider):
    frame = cv2.imread(image_path)
    if frame is None:
        raise ValueError("Could not load image.")

    mask = segmenter.predict_mask(frame)
    laser_scores = decider.laser_scan(mask)
    decision = decider.decide(laser_scores)

    # visualization base
    blended = cv2.addWeighted(frame, 0.7, cv2.applyColorMap(mask, cv2.COLORMAP_JET), 0.3, 0)
    blended_copy = blended.copy()

    # grayscale-shaded laser beam regions based on normalized score
    step = decider.width // decider.num_beams
    max_score = max(laser_scores.values()) if laser_scores else 1
    h = decider.height
    region_centers = {}
    for i, region in enumerate(decider.regions):
        x_start = i * step
        x_end = x_start + step
        norm_score = laser_scores[region] / max_score if max_score > 0 else 0
        intensity = int(255 * (1 - norm_score))  # dark = high score, light = low
        overlay = blended_copy.copy()
        cv2.rectangle(overlay, (x_start, h//2), (x_end, h), (intensity, intensity, intensity), -1)
        blended_copy = cv2.addWeighted(overlay, 0.3, blended_copy, 0.7, 0)
        cv2.rectangle(blended_copy, (x_start, h//2), (x_end, h), (100, 100, 100), 1)
        score_text = f"{laser_scores[region]}"
        cv2.putText(blended_copy, score_text, (x_start + 5, h - 10), cv2.FONT_HERSHEY_PLAIN, 1.2, (230, 230, 230), 1)
        region_centers[region] = (x_start + x_end) // 2


    # direction line
    if decision != "stop" and decision in region_centers:
        center_x = region_centers[decision]
        start_point = (decider.width // 2, h)
        end_point = (center_x, h//2)
        cv2.line(blended_copy, start_point, end_point, (255, 255, 255), 2)

    plt.imshow(cv2.cvtColor(blended_copy, cv2.COLOR_BGR2RGB))
    plt.title(f"Motion Decision: {decision}°")
    plt.axis('off')
    plt.show()

    print("Laser scores:", dict(laser_scores))
    print("Final decision:", decision + "°")


if __name__ == "__main__":
    ckpt_path = "unet_ground_plane.pt"

    last_decision_time = 0
    decision_interval = 3  # seconds

    cap = cv2.VideoCapture(0)
    ret, frame = cap.read()
    h, w = frame.shape[:2]

    out = cv2.VideoWriter("output.mp4", cv2.VideoWriter_fourcc(*'mp4v'), 20, (w, h))  # save video
    segmenter = GroundSegmenter(ckpt_path, device='cpu')
    decider = MotionDeciderWithLasers(segmenter, img_width=w, img_height=h, num_beams=7)
    mask = np.zeros((h, w), dtype=np.uint8)  # initialize dummy mask
    laser_scores = {region: 0 for region in decider.regions}  # initialize empty scores
    decision = "waiting..."

    while True:
        ret, frame = cap.read()
        if not ret:
            break

        h, w = frame.shape[:2]
        decider.width = w
        decider.height = h

        current_time = time.time()
        if current_time - last_decision_time >= decision_interval:
            mask = segmenter.predict_mask(frame)
            laser_scores = decider.laser_scan(mask)
            decision = decider.decide(laser_scores)
            last_decision_time = current_time

        blended = cv2.addWeighted(frame, 0.7, cv2.applyColorMap(mask, cv2.COLORMAP_JET), 0.3, 0)
        blended_copy = blended.copy()

        step = decider.width // decider.num_beams
        max_score = max(laser_scores.values()) if laser_scores else 1
        region_centers = {}

        for i, region in enumerate(decider.regions):
            x_start = i * step
            x_end = x_start + step
            norm_score = laser_scores[region] / max_score if max_score > 0 else 0
            intensity = int(255 * (1 - norm_score))
            overlay = blended_copy.copy()
            cv2.rectangle(overlay, (x_start, h//2), (x_end, h), (intensity, intensity, intensity), -1)
            blended_copy = cv2.addWeighted(overlay, 0.3, blended_copy, 0.7, 0)
            cv2.rectangle(blended_copy, (x_start, h//2), (x_end, h), (100, 100, 100), 1)
            score_text = f"{laser_scores[region]}"
            cv2.putText(blended_copy, score_text, (x_start + 5, h - 10), cv2.FONT_HERSHEY_PLAIN, 1.2, (230, 230, 230), 1)
            region_centers[region] = (x_start + x_end) // 2

        if decision != "stop" and decision != "waiting..." and decision in region_centers:
            center_x = region_centers[decision]
            start_point = (decider.width // 2, h)
            end_point = (center_x, h//2)
            cv2.line(blended_copy, start_point, end_point, (255, 255, 255), 2)

        cv2.imshow("Motion Decision", blended_copy)
        out.write(blended_copy)

        if cv2.waitKey(1) & 0xFF == 27:
            break

    out.release()
    cap.release()
    cv2.destroyAllWindows()

File: motion_decision.cpp
#include <torch/torch.h>
#include <ctime>
#include <opencv2/opencv.hpp>
#include <torch/script.h>
#include <iostream>
#include <chrono>
#include <map>
#include <string>

using namespace std;
using namespace cv; 

# ----------------- Binary Ground Mask Calculation -----------------

class GroundSegmenter {
    public:
        GroundSegmenter(const string& model_path) {
            module = torch::jit::load(model_path);
            module.eval();
        }
    
        pair<Mat, Mat> predict_mask(const Mat& img_bgr) {
            Mat resized, rgb;
            resize(img_bgr, resized, Size(320, 320));
            cvtColor(resized, rgb, COLOR_BGR2RGB);
        
            torch::Tensor tensor = torch::from_blob(rgb.data, {1, 320, 320, 3}, torch::kByte);
            tensor = tensor.permute({0, 3, 1, 2}).to(torch::kFloat) / 255.0;
        
            torch::NoGradGuard no_grad;
            torch::Tensor output = module.forward({tensor}).toTensor();
            output = output.detach().cpu();
            if (output.dim() == 4) output = output.squeeze(0);
            if (output.dim() == 3) output = output.squeeze(0);
        
            Mat prob_map(320, 320, CV_32F);
            for (int i = 0; i < 320; ++i) {
                for (int j = 0; j < 320; ++j) {
                    prob_map.at<float>(i, j) = torch::sigmoid(output[i][j]).item<float>();
                }
            }
        
            Mat mask(320, 320, CV_8UC1);
            for (int i = 0; i < 320; ++i) {
                for (int j = 0; j < 320; ++j) {
                    mask.at<uchar>(i, j) = prob_map.at<float>(i, j) > 0.5f ? 255 : 0;
                }
            }
        
            resize(mask, mask, img_bgr.size());
            resize(prob_map, prob_map, img_bgr.size(), 0, 0, INTER_NEAREST);
        
            return {mask, prob_map};
        }        
    
    private:
        torch::jit::script::Module module;
    };
    
# ----------------- Motion Decision -----------------

    class MotionDeciderWithLasers {
    public:
        MotionDeciderWithLasers(int width, int height, int num_beams = 10) 
            : width(width), height(height), num_beams(num_beams) {
            int angle_step = 180 / (num_beams - 1);
            for (int i = 0; i < num_beams; ++i)
                regions.push_back(to_string(-90 + i * angle_step));
        }
    
        map<string, int> laser_scan(const Mat& mask) {
            map<string, int> laser_scores;
            int step = width / num_beams;
            for (int i = 0; i < num_beams; ++i) {
                int x_start = i * step;
                int x_end = x_start + step;
                Rect region_rect(x_start, height / 2, step, height / 2);
                Mat region = mask(region_rect);
                laser_scores[regions[i]] = countNonZero(region);
            }
            return laser_scores;
        }
    
        string decide(const map<string, int>& scores, float threshold_ratio = 0.05) {
            int total = 0;
            for (const auto& [region, val] : scores)
                total += val;
            if (total == 0) return "stop";
    
            auto best = max_element(scores.begin(), scores.end(),
                [](const auto& a, const auto& b) { return a.second < b.second; });
            if (best->second < total * threshold_ratio)
                return "stop";
            return best->first;
        }
    
        int width, height;
        int num_beams;
        vector<string> regions;
    };

# ----------------- Main Class -----------------
    
    int main() {
        string model_path = "/Users/virginiaceccatelli/Documents/vision_control/unet_ground_plane.pt";
        GroundSegmenter segmenter(model_path);
    
        VideoCapture cap(0);
        if (!cap.isOpened()) {
            cerr << "Camera failed to open" << endl;
            return -1;
        }
    
        Mat frame;
        cap >> frame;
        int w = frame.cols;
        int h = frame.rows;

	int fps = 20;
	VideoWriter out("output_video.mp4", VideoWriter::fourcc('m', 'p', '4', 'v'), fps, Size(w, h));
	VideoWriter heatmap_out("heatmap_output.mp4", VideoWriter::fourcc('m', 'p', '4', 'v'), fps, Size(w, h));

    if (!out.isOpened()) {
	    cerr << "Failed to open video writer" << endl;
	    return -1;
	}
    
        MotionDeciderWithLasers decider(w, h);
        auto last_time = chrono::steady_clock::now();
    
        Mat mask; // no prefilled image
        Mat prob_map; 
	map<string, int> laser_scores;
        string decision = "waiting...";
    
        while (true) {
            cap >> frame;
            if (frame.empty()) break;
    
            auto now = chrono::steady_clock::now();
            float elapsed = chrono::duration_cast<chrono::seconds>(now - last_time).count();
    
            if (elapsed >= 3) {
                tie(mask, prob_map) = segmenter.predict_mask(frame);
                laser_scores = decider.laser_scan(mask);
                decision = decider.decide(laser_scores);
                last_time = now;
            }
    
            if (decision != "waiting...") {
                // Visual Overlay Simple Ground Mask 
                Mat color_mask, blended, overlay;
                applyColorMap(mask, color_mask, COLORMAP_JET);
                addWeighted(frame, 0.7, color_mask, 0.3, 0, blended);
                blended.copyTo(overlay);
    
                int step = decider.width / decider.num_beams;
                int max_score = 1;
                for (const auto& [_, val] : laser_scores)
                    max_score = max(max_score, val);
    
                map<string, int> region_centers;
    
                for (int i = 0; i < decider.num_beams; ++i) {
                    string region = decider.regions[i];
                    int x_start = i * step;
                    int x_end = x_start + step;
                    int val = laser_scores[region];
                    float norm_score = static_cast<float>(val) / max_score;
                    int intensity = static_cast<int>(255 * (1.0f - norm_score));
    
                    rectangle(overlay, Point(x_start, h / 2), Point(x_end, h),
                              Scalar(intensity, intensity, intensity), FILLED);
    
                    region_centers[region] = (x_start + x_end) / 2;
                }
    
                addWeighted(overlay, 0.2, blended, 0.8, 0, blended);
    
                for (int i = 0; i < decider.num_beams; ++i) {
                    int x_start = i * step;
                    int x_end = x_start + step;
                    rectangle(blended, Point(x_start, h / 2), Point(x_end, h), Scalar(100, 100, 100), 1);
    
                    string region = decider.regions[i];
                    int val = laser_scores[region];
                    Point text_pos(x_start + 5, h - 10);
                    putText(blended, to_string(val), text_pos + Point(1, 1),
                            FONT_HERSHEY_PLAIN, 1.3, Scalar(0, 0, 0), 2); // shadow
                    putText(blended, to_string(val), text_pos,
                            FONT_HERSHEY_PLAIN, 1.3, Scalar(255, 255, 255), 1); // text
                }
    
                if (decision != "stop" && region_centers.count(decision)) {
                    int cx = region_centers[decision];
                    Point start(w / 2, h);
                    Point end(cx, h / 2);
                    line(blended, start, end, Scalar(255, 255, 255), 2);
                }
    
                Point text_pos(10, 30);
                putText(blended, "Decision: " + decision + " degrees", text_pos, FONT_HERSHEY_SIMPLEX, 1.0, Scalar(255, 255, 255), 2);

                // Visual Overlay Heatmask
                Mat prob_map_uint8, heatmap, blended_heatmap;
                prob_map.convertTo(prob_map_uint8, CV_8UC1, 255.0);
                applyColorMap(prob_map_uint8, heatmap, COLORMAP_JET);
                addWeighted(frame, 0.6, heatmap, 0.4, 0, blended_heatmap);

                putText(blended_heatmap, "Decision: " + decision + " degrees", Point(10, 30),
                        FONT_HERSHEY_SIMPLEX, 1.0, Scalar(255, 255, 255), 2);
                imshow("Probability Heatmap", blended_heatmap);
                heatmap_out.write(blended_heatmap);

                imshow("Motion Decision Overlay", blended);
		        out.write(blended);
            } else {
                imshow("Motion Decision Overlay", frame); // show raw frame until prediction
            }
    
            if (waitKey(1) == 27) break; // ESC to quit
        }
    
        cap.release();
        out.release();
        heatmap_out.release();
        destroyAllWindows();
        return 0;
    }    
